### DISCUSSION TITLE
Действие при нажатии на кнопку в почте

### DISCUSSION THREAD

**Discussion by Vadimgrig2012** [totumonline/totum-ru-issues-and-discussions#275] [Как это сделать?] [answered]
Всем добрый день. Прошу вашего совета. При отправке сообщения на почту из системы TOTUM необходимо передать в тело письма кнопку (ну или ссылку), при нажатии на которую будет выполнятся какое-либо действие в системе по определенным условиям. Возможна ли реализация данной задачи? Может уже кто-нибудь сталкивался с этим?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Итак — это работает в самой последней версии MIT и PRO (4.8.50.1 and later).

Генерируем ссылку используя `row` с ключами `key` и `value`:

+ `key` — это наш секретный ключ, который мы будем проверять.

+ `value` — данные для действия, которое мы будем выполнять.

Делаем не row тотума, а строку в json сразу функцией `jsonCreate` (она же может и тотум-данные превращать в json).

Что бы это все было защищено мы шифруем функцией `strEncrypt`.

Что бы это все пролезло в `get` нам нужно еще и заурленкодить — `strUrlEncode`.

Дальше я здесь использую шаблон `html` в котором заменяю `get`-параметр на получившийся у меня шифр в `url-encode`.


<img width="1289" alt="Снимок экрана 2022-11-25 в 16 05 06" src="https://user-images.githubusercontent.com/55755565/203992865-91aa783b-abaa-4e6c-b189-3bd67432e1cf.png">

```
= : linkToDataHtml(title: "generated link"; html: $replace)

replace: strReplace(str: $html; from: "{param}"; to: $url)

	url: strUrlEncode(str: $param)
	param: strEncrypt(str: $data)
		data: jsonCreate(field: "key" = "my_secret_key"; field: "value" = "my_value")

    ```html:html
    <a href="https://demoen.totum.online/Remotes/add_by_link?crypt={param}">some generated link</a>
    ```
```

Это и можно вставить в письмо. Только нам нужно еще создать `remotes`, который будет на это отвечать:

Делаем `remotes`, в моем случае с именем `add_by_link`:

```
=: if(condition: $extract[key] = "my_secret_key"; then: $insert)

	insert: insert(table: 'some_table'; field: 'field_with_value' = $extract[value]; inserts: "inserts"; log: true)

~extract: jsonExtract(text: $decrypt)
	decrypt: strDecrypt(str: $#get[crypt])
```

Не забудте назначить пользователя от которого он будет выполнятся и включить его.

В нем мы не делаем `url-decode` тк сервер `url`-кодирование расшифровывает сам автоматически и мы получаем шифрованную строку в `$#get[crypt]`.

Дальше нам надо расшифровать, что у нас там было передано в `crypt` и превратить строку `json` в `row` тотума функцией `jsonExtract` (он фиксируется, что бы несколько раз его не выполнять).

Теперь мы сравниваем ключ и если проверка проходит выполняем действие.

> Если супер-секретное действие, то можно добавить еще один уровень с таблицей, в которой будут хранится временные ключи!

> Ключ шифрования для strEncrypt и strDecrypt находится в файле Crypt.key в корне установки — если его нет, то надо создать с произвольной строкой шифрования. Ключ для шифрования и расшифровки должен быть одинаков.
